lib/multi.c

```diff
@@ -2418,6 +2435,17 @@ static CURLMcode multi_runsingle(struct Curl_multi *multi,
     default:
       return CURLM_INTERNAL_ERROR;
     }
+
+    if(data->conn &&
+       data->mstate >= MSTATE_CONNECT &&
+       data->mstate <= MSTATE_DO &&
+       rc != CURLM_CALL_MULTI_PERFORM &&
+       multi_ischanged(multi, false)) {
```

Comment:
Good question! I agree with your assessment here, and I've changed this to a `!multi_ischanged(multi, false)` because I think it was supposed to match the opposite of the conditions to exit the while loop based on the comment right below.

I've also expanded the comment to explain the last two checks in this check.